firestore-tos-extension/functions/src/index.ts

```
@@ -0,0 +1,145 @@
+
+export const acceptTerms = functions.handler.https.onCall(
+  async (data, context) => {
+    // Checking that the user is authenticated.
+    if (!context.auth) {
+      return;
+    }
+
+    /** check if tosId has been provided  */
+    if (!data.tosId) {
+      console.warn("No tosId provided");
+      return;
+    }
+
+    /** check tos documents */
+    const tosDoc = await db
+      .collection(config.collectionPath)
+      .doc("agreements")
+      .collection("tos")
+      .doc(data.tosId)
+      .get();
+
+    /** Return if no agreement exists  */
+    if (!tosDoc || !tosDoc.exists) {
+      console.warn("tosDoc does not exist");
+      return;
+    }
+
+    /** Find current acknowledgements */
+    const acknowledgements = await db
+      .collection(config.collectionPath)
+      .doc("acknowledgements")
+      .collection(context.auth.uid)
+      .doc(`${process.env.EXT_INSTANCE_ID}`)
+      .get()
+      .then(($) => $.data() || {});
+
+    acknowledgements[data.tosId] = {
+      ...tosDoc.data(),
+      acceptanceDate: admin.firestore.FieldValue.serverTimestamp(),
+    };
+
+    /** Add Acceptance cliam to user document list */
+    await db
+      .collection(config.collectionPath)
+      .doc("acknowledgements")
+      .collection(context.auth.uid)
+      .doc(`${process.env.EXT_INSTANCE_ID}`)
+      .set({ ...(acknowledgements || {}) });
+
+    /** Set claims on user and return */
+    const claims = {};
+    claims[process.env.EXT_INSTANCE_ID] = acknowledgements;
+    return auth.setCustomUserClaims(context.auth.uid, { ...claims });
+  }
+);
+
+export const createTerms = functions.handler.https.onCall(
+  async (data, context) => {
+    // Checking that the user is authenticated.
+    if (!context.auth) {
+      return;
+    }
+
+    /** check if tos data has been provided  */
+    if (!data) {
+      return;
+    }
+
+    /** check if tos data has been provided  */
+    if (!data.noticeType) {
+      return { error: "No noticeType provided" };
```

This doesn't seem right, it should throw an error?

---

I did originally consider https://firebase.google.com/docs/reference/functions/common_providers_https.httpserror#constructor.

Testing proved difficult for catching the error. I'll take another look at improving error responses.

---

The client callable should throw an error if it throws within here. This is technically returning a successful value with an object.

---

Error handling updated.